add lexer support for LoomScript

ellemenno opened the pull request:
LoomScript is the scripting language used by the Loom SDK, a native mobile app and game framework (https://github.com/LoomSDK/LoomSDK). It is largely similar to ActionScript3, with additional C#-ish features.

This pull request refactors a punctuation rule for colons (`:`) in the ActionScript lexer to address a highlighting error that wasn't exercised in the example (e.g. `a -= b;`), then extends the as3 lexer to implement LoomScript highlighting.

ellemenno committed:
correct bug in as3 lexer

`:=` is not an as3 operator, `:` should just be punctuation

ellemenno committed:
formatting tweak; no change except to make single lines

ellemenno committed:
add language reference url to as3 lexer description

ellemenno committed:
remove confusing colon from rackup message to stdout

ellemenno committed:
add lexer support for LoomScript

jneen commented on lib/rouge/lexers/loomscript.rb:
```diff
@@ -0,0 +1,80 @@
+        )
+
+        # builtin types
+        @builtins ||= %w(
```
Further uses of `||=` won't add here, they're a no-op, unfortunately. It desugars to

```
@builtins = @builtins || %w(...)
```

which always is a no-op because `@builtins` is set above.

If you really want comments in-between, what you probably want is:

```
@builtins ||= %w(
) + 

# comment
%w(
...
) +

# comment
%w(
...
)
```

jneen commented:
Also, could you add that snippet to the ActionScript example file? That way we can be sure not to break it in the future.

ellemenno committed:
improve code samples

ellemenno committed:
consistently use %r for regex strings

ellemenno committed:
add token type assertion helpers to lexing.rb

ellemenno committed:
add spec file for actionscript

ellemenno committed:
add spec tests for loomscript where different from actionscript

ellemenno committed:
support doc comment lexing in actionscript and loomscript

ellemenno committed:
correct keywords and names in actionscript and loomscript

also add support for Keyword.Type and Name.Attributes

ellemenno committed:
remove use of tilde heredocs to gain compatibility with ruby 1.9.3

the `<<~` operator is available in ruby 2.3 and later

ellemenno commented:
@jneen Thanks for reviewing; appreciate the catch on the builtins.

I've corrected that issue and added the `-=` snippet to the actionscript example.

I've also added spec tests for both the base lexer (Actionscript) and this new one (LoomScript). The pull request has been rebased on current master.

The new tests all pass using:

```
files=./spec/lexers/*script_spec.rb rake spec
```

When running the full suite, there are errors reported in the `shell_spec` tests. These exist on master without my changes (but these are the errors Travis is reporting above):

```
files=./spec/lexers/shell_spec.rb rake spec
```

ellemenno committed:
ensure the parent lexer definition is loaded prior to referencing it

ellemenno commented:
@jneen changes made and all checks passing

ellemenno committed:
address rubocop warning

ellemenno commented:
@gfx Rebased on current master & passing rubocop checks.
